chemprop/v2/models/multi.py

```diff
@@ -16,24 +16,22 @@ def __init__(
         self,
         message_passing: MulticomponentMessagePassing,
         agg: Aggregation,
-        ffn: nn.Sequential,
-        transform: OutputTransform,
-        loss_fn: LossFunction,
-        metrics: Iterable[Metric],
-        task_weights: Tensor | None = None,
+        readout: Readout,
+        batch_norm: bool = True,
+        metrics: Iterable[Metric] | None = None,
+        w_t: Tensor | None = None,
         warmup_epochs: int = 2,
         init_lr: float = 1e-4,
         max_lr: float = 1e-3,
         final_lr: float = 1e-4,
     ):
         super().__init__(
-            message_passing,
+            message_passing, # MPNN doesn't take MulticomponentMessagePassing?
```

Comment:
@davidegraff Thanks for the comment at https://github.com/chemprop/chemprop/pull/500#discussion_r1367063843! I will move the discussion over to this new PR. Currently, when I try making a MulticomponentMPNN with the minimal working example below, I ran into this error. I tried making `MulticomponentMessagePassing` inherent from `HasHParams` but the same error proceeds. I would like your inputs on resolving this.

```
from chemprop.v2.models import modules, metrics, multi

n_components = 2

mcmp = modules.MulticomponentMessagePassing(
    blocks=[modules.BondMessageBlock() for i in range(n_components)],
    n_components=n_components,
)

# %%
agg = modules.MeanAggregation()

# %%
ffn = modules.RegressionFFN(
)

# %%
metric_list = [metrics.RMSEMetric(), metrics.MAEMetric()]

# %%
mcmpnn = multi.MulticomponentMPNN(
    mcmp,
    agg,
    ffn,
    metrics=metric_list,
)

mcmpnn
```

```
Exception has occurred: AttributeError
'MulticomponentMessagePassing' object has no attribute 'hparams'
  File "/Users/hwpang/Software/chemprop/chemprop/v2/models/model.py", line 79, in __init__
    "message_passing": message_passing.hparams,
                       ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hwpang/Software/chemprop/chemprop/v2/models/multi.py", line 28, in __init__
    super().__init__(
  File "/Users/hwpang/Software/chemprop/examples/debug.py", line 104, in <module>
    mcmpnn = multi.MulticomponentMPNN(
             ^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'MulticomponentMessagePassing' object has no attribute 'hparams'
```

Reply:
ah yeah another bug... Essentially, each module in chemprop implements the `HasHParams` protocol, which is to say that each module has an attribute `hparams` that is a dictionary with the key `"cls"` that stores the module's class and a bunch of other keys equal to the module hyperparameters, i.e.,
```python
mp = BondMessageBlock()
print(mp)
# BondMessageBlock(
#     [...]
# )
print(type(mp.hparams))
# dict
print(mp.hparams)
{'cls': type<BondMessageBlock>, 'd_v': 133, 'd_e': 147, 'd_h': 300, ...}
mp_cls = mp.hparams.pop('cls')
mp_new = mp_cls(**mp.hparams)
print(mp_new)
# BondMessageBlock(
#     [...]
# )
```
This is necessary to get model checkpointing to work efficiently within the PyTorch Lightning framework because our `MPNN` is itself composed of `nn.Module`s (rather than simple hyperparameters).

Looking at the `MessagePassingBlockBase` class there are two things to note:
1. The class signature: `class MessagePassingBlockBase(MessagePassingBlock, HyperparametersMixin)`
2. a call to `self.save_hyperparameters()` right after `super().__init__()`

(1) is a mixin class from PyTorch lightning which allows us to use the `save_hyperparameters()` method. This method stores all of our input arguments to, you guessed it, the `hparams` attribute of the object. Right below that call, there's also this line:

    self.hparams["cls"] = self.__class__

This updates the `hparams` attribute such that the class satisfies the `HasHParams` protocol I mentioned above.

All of this is to say, that I think you can fix this error by adding the following lines to `message_passing.multi.py`:

```python
class MulticomponentMessagePassing(nn.Module, HasHParams):
    """
    [...]
    """

    def __init__(
        self, blocks: Sequence[MessagePassingBlockBase], n_components: int, shared: bool = False
    ):
        super().__init__()
        self.hparams = {
            "cls": self.__class__,
            "blocks": [block.hparams for block in blocks],
            "n_components": n_components,
            "shared": shared
        }
```
and then we'd also need to override `MulticomponentMPNN.load_from_checkpoint`:
```python
    @classmethod
    def load_from_checkpoint(
        cls, checkpoint_path, map_location=None, hparams_file=None, strict=True, **kwargs
    ) -> MPNN:
        hparams = torch.load(checkpoint_path)["hyper_parameters"]

        # construct the `agg` and `readout` modules from their hparam dicts like before
        # and store them into `kwargs`
        agg_hparams = hparams["agg"]
        agg = agg_hparams.pop("cls")(**agg_hparams)
        kwargs["agg"] = agg

        readout_hparams = hparams["readout"]
        readout = readout_hparams.pop("cls")(**readout_hparams)
        kwargs["readout"] = readout

        # iteratively construct the input message blocks from their corresponding hparams
        # we do this because the `MulticomponentMessagePassing.hparams` attribute is a nested
        # dictionary of hparams 
        mp_hparams = hparams["message_passing"]
        mp_hparams["blocks"] = [
            block_haparams.pop("cls")(**block_hparams) for block_hparams in mp_hparams["blocks"]
        ]
        message_passing = mp_hparams.pop("cls")(**mp_hparams)
        kwargs["message_passing"] = message_passing

        return super().load_from_checkpoint(
            checkpoint_path, map_location, hparams_file, strict, **kwargs
        )
```
One thing to note in this implementation is that we'll need to directly call the `LightningModule.load_from_checkpoint` rather than `MPNN.load_from_checkpoing` (the `super()` call implicitly does the latter). There is another option that I can think of, but I'll omit it for now to avoid possible confusion.